src/snowflake/snowpark/ml/transformer.py

```diff
@@ -0,0 +1,395 @@
+    def input_cols(self, value: List[str]) -> None:
+        self._input_cols = value
+        self._states_table_cols = (
+            [f"states_{c}" for c in self._input_cols] if self._input_cols else None
+        )
+        self._encoder_table_cols = (
+            [f"states_{c}_encoder" for c in self._input_cols]
+            if self._input_cols
+            else None
+        )
+
+    def fit(self, df: DataFrame) -> "OneHotEncoder":
+        if not self._input_cols:
+            raise ValueError("Input columns can not be None")
+
+        encoder_fit(self, df, "onehot")
+        return self
+
+    def transform(
+        self,
+        df: DataFrame,
+        sparse: bool = False,
+    ) -> DataFrame:
+        if self._output_cols is None:
+            raise ValueError("Output column can not be empty")
+        elif len(self._output_cols) == 1:
+            self._merge = True
+        elif len(self._input_cols) == len(self._output_cols):
+            self._merge = False
+        else:
+            raise ValueError(
+                "The number of output column and input column does not match"
+            )
+
+        original_columns = df.columns
+        encoder_structure = [] if not sparse else array_construct()
+        encoder_offset = lit(0)
+        df_count_table = df._session.table(self._encoder_count_table)
+        df = df.join(df_count_table)
+
+        for fc, ec, (c, table_name) in zip(
+            self._states_table_cols,
+            self._encoder_table_cols,
+            self._category_table_name.items(),
+        ):
+            category_table = df._session.table(table_name)
+            # if sparse is set to true, covert category_table so that it store onehot array
+            if sparse:
+                df_sparse_array = category_table.with_column("sparse", lit(0)).select(
+                    [array_agg(col("sparse")).as_(f"{fc}_sparse")]
+                )
+                df_sparse_array = category_table.join(df_sparse_array).join(
+                    df_count_table
+                )
+                category_table = df_sparse_array.select(
+                    [
+                        col(fc),
+                        array_slice(
+                            array_insert(col(f"{fc}_sparse"), col(ec), lit(1)),
+                            lit(0),
+                            col(f"{fc}_count"),
+                        ).as_(ec),
+                    ]
+                )
+                df = category_table.join(
+                    df, df[c] == category_table[fc], join_type="right"
+                ).join(df_sparse_array.select([col(f"{fc}_sparse")]).limit(1))
+            else:
+                df = category_table.join(
+                    df, df[c] == category_table[fc], join_type="right"
+                )
+            if self._merge:
+                if not sparse:
+
+                    encoder_structure.append(
+                        iff(col(ec).is_null(), col(ec), col(ec) + encoder_offset)
```

Comment:
add a comment saying offset is used to shift from one category to another category?